src/main/java/Duke.java

```diff
@@ -1,10 +1,150 @@
+                        t.markAsDone();
+                        tasks.add(i, t);
+                        System.out.println("Nice! I've marked this task as done:");
+                        System.out.println(t);
+                        break;
+                    case "delete":
+                        int j = Integer.parseInt(command[1]) - 1;
+                        if (j+1 > tasks.size() || j+1 < 0) throw new DukeException("\u2639 OOPS!!! This task does not exist.");
+                        Task tsk = tasks.remove(j);
+                        System.out.println("Noted. I've removed this task:");
+                        System.out.println(tsk);
+                        System.out.print("Now you have " + tasks.size());
+                        if (tasks.size() == 1) System.out.println(" task in the list.");
+                        else System.out.println(" tasks in the list.");
+                        break;
+                    case "todo":
+                        if (command.length < 2)
+                            throw new DukeException("\u2639 OOPS!!! The description of a todo cannot be empty.");
+                        Task todo = new Todo(command[1]);
+                        tasks.add(todo);
+                        printAddedTask(todo, tasks.size());
+                        break;
+                    case "deadline":
+                        if (command.length < 2)
+                            throw new DukeException("\u2639 OOPS!!! The description of a deadline cannot be empty.");
+                        String[] detailsD = command[1].split(" /by ");
+                        if (detailsD.length < 2)
+                            throw new DukeException("\u2639 OOPS!!! Deadline missing date and time.");
+                        Task deadline = new Deadline(detailsD[0], detailsD[1]);
+                        tasks.add(deadline);
+                        printAddedTask(deadline, tasks.size());
+                        break;
+                    case "event":
+                        if (command.length < 2)
+                            throw new DukeException("\u2639 OOPS!!! The description of an event cannot be empty.");
+                        String[] detailsE = command[1].split(" /at ");
+                        if (detailsE.length < 2)
+                            throw new DukeException("\u2639 OOPS!!! Event missing start and end time.");
+                        Task event = new Event(detailsE[0], detailsE[1]);
+                        tasks.add(event);
+                        printAddedTask(event, tasks.size());
+                        break;
+                    default:
+                        hasChanged = false;
+                        throw new DukeException("\u2639 OOPS!!! I'm sorry, but I don't know what that means :-(");
+                }
+            } catch (DukeException ex) {
+                System.out.println(ex.getMessage());
+            } finally {
+                if (hasChanged) {
```

Comment:
The current implementation of saving your tasks after every command is fine for now. However, this process may take a long time when the to-do list gets long. Perhaps save only when the 'bye' command is entered?

Reply:
I think @shiyao821, your suggestion is definitely the more efficient one in terms of the number of times you write to your file in memory, but I guess as @larrylawl pointed out, the website seems to indicate otherwise. Right now, with all the GUI also set up, user can just hit the close button on the top right of the interface to exit from the application, and this can be problematic if we only save all changes to the task list when user enters command `bye`. So, I am going with writing the tasks in the file again each time the user changes the task list in any way.